mmdet/datasets/pipelines/copy_paste_aug.py

```diff
@@ -0,0 +1,195 @@
+        composed_mask = composed_mask[..., None]
+        results['img'] = results['copy_paste']['img'] * composed_mask + results['img'] * (1 - composed_mask)
+        results['img'] = results['img'].astype(img_dtype)
+
+    def masks_copy_paste(self, results, paste_objects, composed_mask):
+        results['gt_masks'].masks = np.array([
+            np.where(composed_mask == 1, 0, mask).astype(np.uint8) for mask in results['gt_masks'].masks
+        ])
+        results['gt_masks'].masks = np.concatenate(
+            (results['gt_masks'].masks, results['copy_paste']['gt_masks'].masks[paste_objects]), axis=0)
+        results['gt_bboxes'] = np.concatenate(
+            (results['gt_bboxes'], results['copy_paste']['gt_bboxes'][paste_objects]), axis=0)
+        self.extract_bboxes(results)
+
+    def concatenate_labels(self, results, paste_objects=None):
+        results['gt_labels'] = np.concatenate(
+            (results['gt_labels'], results['copy_paste']['gt_labels'][paste_objects]), axis=0)
+        if not self.filter_ignored:
+            results['gt_bboxes_ignore'] = np.concatenate(
+                (results['gt_bboxes_ignore'], results['copy_paste']['gt_bboxes_ignore']), axis=0)
+
+    @staticmethod
+    def _filter(results, inds):
+        results['gt_masks'].masks = results['gt_masks'].masks[inds]
+        results['gt_bboxes'] = results['gt_bboxes'][inds]
+        results['gt_labels'] = results['gt_labels'][inds]
+        if len(results['gt_masks'].masks.shape) == 2:
+            results['gt_masks'].masks = np.expand_dims(results['gt_masks'].masks, axis=0)
+        if len(results['gt_bboxes'].shape) == 1:
+            results['gt_bboxes'] = np.expand_dims(results['gt_bboxes'], axis=0)
+        if len(results['gt_labels'].shape) == 0:
+            results['gt_labels'] = np.expand_dims(results['gt_labels'], axis=0)
+
+    def filter_empty(self, results, min_size=5):
+        w = results['gt_bboxes'][:, 2] - results['gt_bboxes'][:, 0]
+        h = results['gt_bboxes'][:, 3] - results['gt_bboxes'][:, 1]
+        # Filter too small objects
+        inds = np.argwhere((w > min_size) & (h > min_size)).squeeze()
+        self._filter(results, inds)
+        # Filter objects with empty mask
+        inds = np.argwhere(results['gt_masks'].areas > 0).squeeze()
+        self._filter(results, inds)
+
+    def filter_ignored_objects(self, results, k=0.9):
+        if len(results['gt_bboxes_ignore']) == 0:
+            return
+        inds = []
+        for i, bbox in enumerate(results['gt_bboxes']):
+            c = 0
+            for ignore in results['gt_bboxes_ignore']:
+                x0, x1 = max(bbox[0], ignore[0]), min(bbox[2], ignore[2])
+                y0, y1 = max(bbox[1], ignore[1]), min(bbox[3], ignore[3])
+                s_intersection = (x1 - x0) * (y1 - y0)
+                s_bbox = (bbox[2] - bbox[0]) * (bbox[3] - bbox[1])
+                if s_intersection > k * s_bbox:
+                    c += 1
```

Comment:
Is it safe to break the loop at this point?

Reply:
Yes. Refactored a bit.